docs/components/api.md

```diff
@@ -23,28 +23,28 @@ The whole API provided by stencil can be condensed in a set of decorators, lifec
 Decorators are a pure compiler-time construction used by stencil to collect all the metadata about a component, the properties, attributes and methods it might expose, the events it might emit or even the associated stylesheets.
 Once all the metadata has been collected, all the decorators are removed from the output, so they don't incur any runtime overhead.
 
-- [@Component()](component#component-decorator) declares a new web component
-- [@Prop()](properties#prop-decorator) declares an exposed property/attribute
-- [@State()](state#the-state-decorator-state) declares an internal state of the component
-- [@Watch()](reactive-data#watch-decorator) declares a hook that runs when a property or state changes
-- [@Element()](host-element#element-decorator) declares a reference to the host element
-- [@Method()](methods#method-decorator) declares an exposed public method
-- [@Event()](events#event-decorator) declares a DOM event the component might emit
-- [@Listen()](events#listen-decorator) listens for DOM events
+- [@Component()](./component#component-decorator) declares a new web component
+- [@Prop()](./properties#prop-decorator) declares an exposed property/attribute
```

Comment:
It may not, I think it will unfortunately only check anchors within the same file. What I ended up doing was basically converting all of these links to be relative links that start with `./`. This works with Docusaurus properly, and it will check that the link resolves to a page, but I don't think that Docusaurus verifies that specific anchors within a linked page exist.

The markdown link checker, by contrast, checks that links either resolve to anchor links *within the same file* (if they look like `[foo](#foo)` _or_ that they'll return a 200 when it makes a `GET` request to them (more or less). So basically it won't know what to do with the links that Docusaurus renders as internal links.

So anyhow, what I did to get Docusaurus' link-checking and the markdown link checker's link checking to work somewhat well together is converting Docusaurus links to use the relative path and then ignoring those links in the markdown link checker.

However, one thing I just tried is using the alternate file path linking supported in Docusaurus, like

```md
[link](./foo.md#foo)
```

Both Docusaurus and the markdown link checker will check this and report that yes, this is pointing to a file in the right spot, but neither will check that the anchor tag `foo` is present in that file. Boo!

There's an issue on the Docusaurus repo here about this: https://github.com/facebook/docusaurus/issues/3321

One suggestion in there is to convert to file-path links (`[](./foo.md)`) and then use a project like this one: https://github.com/remarkjs/remark-validate-links which might provide a better way to validate all the _internal_ links between files. It looks like it checks that anchor tags are present in any linked file.

_That_ option of course doesn't cover _external_ links.

So my takeaway here is that there's not a perfect solution here. I think I'd be in favor of the following:

- adding the markdown link checker, along the lines of how it's set up here where it's constrained to only checking external URLs
- refactoring internal links to use the filepath-based linking
- adding `remark-validate-links` to check that all internal links + anchor tags are valid (this would make it basically an extra layer of validation on top of what will already get checked by `docusaurus build`).

Thoughts?

Reply:
The remark plugin seems pretty cool as long as all of the tools can play together nicely. Writing relative file paths might be a bit tedious, but I feel like we had to do that sometimes in the old docs too so nbd. I don't really have a strong opinion here for next steps, but think this is a step in the right direction and probably better than the nothing we have today 😬

Reply:
This sounds like a plan to me. The only thing I'm hesitant on is converting internal links to use the filepath-based linking. I think this could make some maintenance a little tricky if we rename a file, but perhaps that's the cost you have to pay if you want anchor tag validation at the moment. I think I'm willing to pay that cost if it means automated tests. Does `remark-validate-links` flag you if you don't use file-path links? My other concern here is education across teams contributing to the site. If `remark-validate-links` doesn't flag non-file-path links, maybe the compromise here is to just move forward with `mlc` only here, and see how that works for us (taking that usage period as a time to better understand our needs)

Reply:
So I was just able to confirm that `remark-validate-links` _does_ show a warning if you don't use file-path-based links. If you have, for instance, something like this:

```md
| `single-export-module`        | All component and custom element definition helper functions will be exported from the `index.js` file in the output directory. This file can be used as the root module when distributing your component library, see [below](../custom-elements#distributing-custom-elements) for more details. |
```

it will say this:

<img width="1277" alt="Screenshot 2023-02-09 at 10 33 40 AM" src="https://user-images.githubusercontent.com/6207644/217858562-4e28ae6d-baa7-4355-98f1-0f205a823514.png">

this can be addressed (in this case) by changing the markdown to this:

```md
| `single-export-module`        | All component and custom element definition helper functions will be exported from the `index.js` file in the output directory. This file can be used as the root module when distributing your component library, see [below](../output-targets/custom-elements.md#distributing-custom-elements) for more details. |
```

A slightly confusing thing with re-writing these to relative-file-path based links is that right now they are effectively relative paths _based on their final url_, so for instance from `docs/introduction/upgrading-to-stencil-three.md`, which has the `url` set to `url: /docs/upgrading-to-stencil-3` in the frontmatter the relative url to the page rendered based on `docs/output-targets/custom-elements.md` is actually just `../custom-elements` because _that_ file has this in frontmatter: `slug: /custom-elements`

So it's a little bit funky to rewrite them, you can sometimes just put `.md` at the end of what's there already, but sometimes you have to figure out which `.md` file the relative URL would normally point to and then find it. Not impossible, but a little funky.

One consequence of that is that right now individual markdown files can be somewhat 'mobile', i.e. they can be moved around without all their relative links being broken. With this change, moving a file would require fixing all of the relative paths to other files (possibly some IDEs can fix such things easily). Anyway, not a major concern because we don't move files around all that much, but worth flagging.

Reply:
I just pushed a change to convert all the internal links over to filepath based ones, fixing any missing anchor tags as well. I used `remark-validate-links` locally to do so, it's quite handy! If you have `[](./foo.md#foo-bar)` and it should actually be `[](./foo.md#foo-bar-baz)` it will do a little 'did you mean.. ?' suggestion (I think based on prefix matching). It's not so many that need to change actually, and it did catch a few broken anchors, so that's nice.

The question then is, if we want to use the `remark-validate-links` thing, should that be installed as a dependency, or installed only as a github action? I'd kind of lean towards installing as a project dependency so we can add a `package.json` script for running it for checking things out locally before pushing.

Reply:
ok so update — that only addressed a small portion of the internal links. There's a lot more!

Reply:
👍 Would you consider this ready to re-review then? I'd like to pull the branch down and do some faux-development with it.

Reply:
feel free to play with it, I am going through and changing over to relative links everywhere -- this requires editing ~400 or so of them and I'm not quite done yet!

Reply:
I'm going to add `remark-validate-links` to dev dependencies and then add running it to the CI as well